gdx/src/com/badlogic/gdx/maps/tiled/BaseTmxMapLoader.java

```diff
@@ -160,6 +160,25 @@ protected TiledMap loadTiledMap (FileHandle tmxFile, P parameter, ImageResolver
 			Element element = root.getChild(i);
 			loadLayer(map, map.getLayers(), element, tmxFile, imageResolver);
 		}
+
+		// update hierarchical parallax scrolling factors
+		// in Tiled the final parallax scrolling factor of a layer is the multiplication of its factor with all its parents
+		// 1) get top level groups
+		final Array<MapGroupLayer> groups = map.getLayers().getByType(MapGroupLayer.class);
```

Comment:
I'm wondering if it's more appropriate to compound the parallax factors in the renderer, rather than in the loader. It just seems easier than the bookkeeping involved in keeping them in sync (given they are mutable).

Reply:
Do you mean to calculate it each render call? I thought about it as well but did not like it because it is a immutable value and only needs to be calculated once in my opinion which makes sense in the loader?
Also, theoretically you can have multiple nestings I think and going up the hierarchy to get all parent factor values for the final value of the leaf layer is most likely worse performance wise.

However, you are right that right now it is a mutable property which should not be the case maybe. But making it a constructor parameter to make it immutable might be tricky because we already need the final value including the group inheritance.

Reply:
i don't think that calculating global parallax factor is so costly. It's about few float multiplications which is nothing compared to layer rendering. I think people will definitely want to have it mutable and they probably don't want to calculate global parallax them-self when changing it for a single layer, so i think it would be better to consider layer parallax property to be "local parallax" and have "global parallax" calculated on the fly at rendering time as suggested by MobiDevelop. What do you think?

Reply:
Sure, I can change that! Just be aware - if I remember correctly it is not just float multiplication. It is also the lookup of those values and if I remember correctly this can be achieved via the parent property of a MapLayer. And it is not limited to just one "jump" because GroupLayers can be nested and therefore multiple "parent.getParallaxFactor" calls can happen each frame.

As you said though it is most likely not that costy and allows then also for programmatic changes to the factor values if the user wants that.

I think it is a good idea 👍

Reply:
Maybe we can have an accumulator that sits at the root and every time you run into a group layer you do apply it's factor (and unapply it after the group). That way you're not looking up parent layers down the chain.

Reply:
In general I like your idea @MobiDevelop but I see the problem that the rendering code seems to be copy&pasted in multiple renderers and therefore we'd need to introduce this accumulator handling in multiple places? Or, refactor it and extract the beginning part of `renderTiledLayer` / `renderImageLayer` which seems to be always the same in every renderer?

However, I don't want that to be part of this PR because it is more effort and riskier and I don't feel comfortable with that, sorry.

----

What I can offer for now is something like this in `MapLayer`:

```Java
/** @return layer's parallax scrolling factor for x-axis including the factor of all its parents */
	public float getGlobalParallaxX () {
		if (parent == null) {
			return parallaxX;
		}

		float acc = 1f;
		MapLayer curr = parent;
		while (curr != null) {
			acc *= curr.getParallaxX();
			curr = curr.parent;
		}
		return parallaxX * acc;
	}
```

And then adjust the renderer lines like:

```Java
// from this
final float layerOffsetX = layer.getRenderOffsetX() * unitScale + viewBounds.x * (layer.getParallaxX() - 1);
final float layerOffsetY = -layer.getRenderOffsetY() * unitScale - viewBounds.y * (layer.getParallaxY() - 1);

// to this
final float layerOffsetX = layer.getRenderOffsetX() * unitScale + viewBounds.x * (layer.getGlobalParallaxX() - 1);
final float layerOffsetY = -layer.getRenderOffsetY() * unitScale - viewBounds.y * (layer.getGlobalParallaxY() - 1);
```

What I personally don't like here is that the parent loop actually needs to be executed twice which is not necessary if we introduce a `Vector2`. Maybe a static Vector2 instance in `MapLayer` to achieve something like this?

```Java
// MapLayer
private static final Vector2 TMP_VEC = new Vector2();
	public Vector2 getGlobalParallaxFactor () {
		TMP_VEC.set(1f, 1f);
		if (parent == null) {
			return TMP_VEC;
		}

		MapLayer curr = parent;
		while (curr != null) {
			TMP_VEC.set(
				TMP_VEC.x * curr.getParallaxX(),
				TMP_VEC.y * curr.getParallaxY()
			);
			curr = curr.parent;
		}
		return TMP_VEC;
	}

// and then in the renderer we can call this only once
final Vector2 globParallax = layer.getGlobalParallaxFactor()
final float layerOffsetX = layer.getRenderOffsetX() * unitScale + viewBounds.x * (layer.getParallaxX() * globParallax.x - 1);
final float layerOffsetY = -layer.getRenderOffsetY() * unitScale - viewBounds.y * (layer.getParallaxY() * globParallax .y - 1);
```

What do you think @MobiDevelop  / @mgsx-dev ?